crates/holochain/src/conductor/conductor/app_connection_auth.rs

```diff
@@ -0,0 +1,239 @@
+use holochain_conductor_api::AppAuthenticationToken;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::time::SystemTime;
+
+use crate::conductor::error::{ConductorError, ConductorResult};
+use holochain_types::prelude::InstalledAppId;
+
+pub struct AppAuthTokenStore {
+    issued_tokens: HashMap<AppAuthenticationToken, TokenMeta>,
+}
+
+impl AppAuthTokenStore {
+    pub fn new() -> Self {
+        Self {
+            issued_tokens: HashMap::new(),
+        }
+    }
+
+    /// Issue a token that can be used to authenticate a connection. The token will only be valid
+    /// for use with the specified `installed_app_id` and will expire after `expiry_seconds`.
+    ///
+    /// If `single_use` is true, the token will be invalidated after the first use, successful or not.
+    pub fn issue_token(
+        &mut self,
+        installed_app_id: InstalledAppId,
+        expiry_seconds: u64,
+        single_use: bool,
+    ) -> (AppAuthenticationToken, Option<SystemTime>) {
+        let token = nanoid::nanoid!().into_bytes();
```

Comment:
nanoid has a bunch of extra overhead to make it human readable. If we're converting it to bytes anyways, we could just use rand to fill a buffer.

Reply:
Ahh good point, done